093.mediawiki

```diff
@@ -0,0 +1,329 @@
+ |}
+
+The vAdditionalTxs field should include at least the coinbase transaction, since it is not possible for it to be in the receiver's mempool.
+
+====CGrapheneSet====
+''CGrapheneSet'' contains all information critical to the Graphene set reconciliation process, absent any block-specific details. It is the only non-standard data structure in ''CGrapheneBlock''.
+
+{|
+ |'''Field Name'''||'''Type'''||'''Size'''||'''Purpose'''
+ |-
+ |ordered||uint8_t||1 byte||1 if order is important in the set and 0 otherwise; other values are reserved for future use
+ |-
+ |nReceiverUniverseItems||uint64_t||8 bytes||The number of transactions in the receiver's mempool
+ |-
+ |encodedRank||<nowiki>vector<unsigned char></nowiki>||ordered * ceil(n log(n))/8 bytes||Order information for items in set
+ |-
+ |setFilter||CBloomFilter||variable||Bloom filter containing items in set
+ |-
+ |setIblt||CIblt||variable||IBLT containing items in set
+ |}
+
+====CBloomFilter====
+''CBloomFilter'' is Bitcoin's standard Bloom filter implementation with one adjustment, described below. It is used as part of the set reconciliation process in ''CGrapheneSet'' to filter out transactions from ''T'' that do not belong to the block.
+
+{|
+ |'''Field Name'''||'''Type'''||'''Size'''||'''Purpose'''
+ |-
+ |vData||<nowiki>vector<unsigned char></nowiki>||variable||Bit array for filter
+ |-
+ |isFull||bool||1 byte||True if every bit in vData is set to 1
+ |-
+ |isEmpty||bool||1 byte||True if every bit in vData is set to 0
+ |-
+ |nHashFuncs||unsigned int||2, 4 bytes||Number of hash functions used in filter
+ |-
+ |nTweak||unsigned int||2, 4, bytes||Additive offset for hash function inputs
+ |-
+ |nFlags||unsigned char||1 byte||Defines behavior for transaction insertion
+ |}
+
+The standard Bloom filter implementation does not always honor the false positive rate (FPR) requested during initialization. This behavior is problematic for Graphene because the FPR determines the number of erroneous transactions removed from IBLT ''I'', which affects the decode rate for ''I''. For this reason, it is required that ''nDesiredSize'' and ''nHashFuncs'' be defined slightly differently than in the standard ''CBloomFilter'' implementation, as follows.
+
+<code>
+nDesiredSize = (unsigned int)(ceil(-1 / LN2SQUARED * nElements * log(nFPRate) / 8));
+</code>
+<br />
+<code>
+nHashFuncs = (unsigned int)max(MIN_N_HASH_FUNC, int(vData.size() * 8 / nElements * LN2));
+</code>
+
+====CIblt====
+''CIblt'' is an IBLT implementation that is part of the ''CGrapheneSet'' reconciliation process. In the absence of a reconciliation failure, ''CIblt'' will generate false positive list ''F'' and missing transaction cheap hashes ''M''.
+
+{|
+ |'''Field Name'''||'''Type'''||'''Size'''||'''Encoding Details'''||'''Purpose'''
+ |-
+ |version||uint64_t||8 bytes||Compact size||Version bits
+ |-
+ |n_hash||uint8_t||1 byte||Standard||Number of hash functions used
```

Comment:
My understanding is the hash functions need to be different functions for multiple functions to be useful.  I'm not sure that n_hash can achieve that definition on its own.

Reply:
That's kinda neat.  I think this needs to be made explicit in the spec.  Specifying the number of hashes is a nice way to keep the message small, but without knowing that the hash functions are implemented this way, it becomes impossible to implement this spec from scratch.

Reply:
@jasonbcox there's a note at the end of the CIblt section: 

> ...it is critical that the IBLTs use the same quantity of hash functions, have the same number of cells, and that hash function i uses the same function and seed for both I and I' ...The i'th hash function is the MurmurHash3 with seed i.

Do you feel that this note is ambiguous or are you saying that it should be made more prominent?

Reply:
Ah.  I don't recall this being in the version I reviewed, but I may have simply missed it.  Thanks.